lib/internal/modules/esm/loader.js

```diff
@@ -373,22 +390,44 @@ class ESMLoader {
       const { compileFunction } = require('vm');
       const preloadInit = compileFunction(
         preload,
-        ['getBuiltin'],
+        ['getBuiltin', 'port', 'setImportMetaCallback'],
         {
           filename: '<preload>',
         }
       );
       const { NativeModule } = require('internal/bootstrap/loaders');
-
-      FunctionPrototypeCall(preloadInit, globalThis, (builtinName) => {
-        if (NativeModule.canBeRequiredByUsers(builtinName)) {
-          return require(builtinName);
+      let finished = false;
+      let replacedImportMetaInitializer = false;
+      let next = this.#importMetaInitializer;
+      try {
+        FunctionPrototypeCall(preloadInit, globalThis, (builtinName) => {
+          if (NativeModule.canBeRequiredByUsers(builtinName)) {
+            return require(builtinName);
+          }
+          throw new ERR_INVALID_ARG_VALUE('builtinName', builtinName);
+        }, insidePreload, (fn) => {
+          if (finished || typeof fn !== 'function') {
+            throw new ERR_INVALID_ARG_TYPE('fn', fn);
+          }
+          replacedImportMetaInitializer = true;
+          const parent = next;
+          next = (meta, context) => {
+            return fn(meta, context, parent);
+          };
+        });
+      } finally {
+        finished = true;
+        if (replacedImportMetaInitializer) {
+          this.#importMetaInitializer = next;
         }
-        throw new ERR_INVALID_ARG_VALUE('builtinName', builtinName);
-      });
+      }
     }
   }
 
+  importMetaInitialize(meta, context) {
+    this.#importMetaInitializer(meta, context);
+  }
```

Comment:
This isn’t creating a new hook, is it? It looks like it’s just moving code that used to be in `translators.js`.

Reply:
depends on what you want to call it, it lets loaders interact with import.meta which they can't prior to this. I wouldn't call it a new hook since it doesn't affect lifecycle of a module being loaded.

Reply:
also note that this function cannot be done via a loader `export` due to living in a different context potentially

Reply:
the "hooks" is a kind of vague term here, where this is able to replace the default logic but not in the same way as `resolve`, `load`, or `globalPreload`. The ability to replace this function is done in the global preload code arguments, but not via `exports` on the loader ESM source text. You can replace parts on `import.meta` like in the example in the docs:  https://github.com/nodejs/node/pull/39240/files#diff-08a5d722c9df06780af1b452e4394133c678b11eee19530ef0bd871f4a8a2c2fR59